src/store/reducers.ts

```diff
@@ -8,50 +11,60 @@ import {
   DELETE_HABIT,
   UPDATE_STREAK,
   UPDATE_CURRENT_USER,
-  REGISTER_USER
-} from "../actions";
+  REGISTER_USER,
+  State,
+  ActionTypes,
+  UserState,
+} from "./types";
 
-const initialState = {
+const initialState: State = {
   initialized: false, // false means the state hasn't been read from localstorage yet
   currentUser: null, // Null means no one is logged in, a string identifies a username
-  userStates: {}
+  userStates: {},
 };
 
 const STATE_KEY = "state";
 
 // Write redux state to local storage
-export const writeStateMiddleware = store => next => action => {
+export const writeStateMiddleware: Middleware<{}, State> = (store) => (
+  next
+) => (action: ActionTypes) => {
   let result = next(action);
   const { initialized, ...state } = store.getState();
   window.localStorage.setItem(STATE_KEY, JSON.stringify(state));
   return result;
 };
 
 // Retrieve Redux state from local storage
-export const readState = () => {
+export const readState = (): State => {
   try {
     const stateJson = window.localStorage.getItem(STATE_KEY);
-    const parsedState = JSON.parse(stateJson);
-    return decodeState(parsedState);
-  } catch {
+    if (stateJson) {
+      // TODO: verify that stateJSON parsing succeeded and is of type StateJSON
```

Comment:
I think this is important. It shouldn't hold up this pull request, but we should do some actual JSON parsing here and verify that we have the value we're asserting we have -- `StateJSON` -- as right now we're basically saying "we promise!" without any verification that we actually have a valid value of this type. I think it's OK to leave this TODO comment in for now but we should loop back around to this.